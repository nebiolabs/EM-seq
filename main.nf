nextflow.enable.dsl=2

/* --------------- *
 * INPUT ARGUMENTS *
 * --------------- */
params.email                     = 'undefined'
params.flowcell                  = 'undefined'
params.path_to_genome_fasta      = 'undefined' // HAS TO BE THE FULL PATH!! 
params.input_glob                = '*_R1.fastq*' // either the .bam or fastq read 1
params.project                   = 'project_undefined'
params.workflow                  = 'EM-seq'
params.outputDir                 = "em-seq_output" 
params.tmp_dir                   = '/tmp'
params.min_mapq                  = 20 // for methylation assessment.
params.max_input_reads           = "all_reads" // default is not downsampling , set to a number to downsample e.g. 1000000 is 500k read pairs
params.downsample_seed           = 42
params.enable_neb_agg            = 'False'

include { alignReads; mergeAndMarkDuplicates; bwa_index; enough_reads; send_email; touchFile }          from './modules/alignment'
include { methylDackel_mbias; methylDackel_extract }                                                    from './modules/methylation'
include { gc_bias; idx_stats; flag_stats; fastqc; insert_size_metrics; picard_metrics; tasmanian }      from './modules/compute_statistics'
include { aggregate_emseq; multiqc }                                                                    from './modules/aggregation'


// custom fx
def replaceReadNumber(inputString) {
    return inputString.replace('_R1.', '_R2.')
                      .replace('_1.fastq', '_2.fastq')
                      .replace('_R1_', '_R2_')
                      .replace('.R1.', '.R2.')
                      .replace('_1.', '_2.')
                      .replace('.1.fastq', '.2.fastq')
}

// detect bam or fastq (or fastq.gz)
def detectFileType(file) {
    def file_str = file.toString()
    if (file_str.endsWith('.bam')) {
        return 'bam'
    } else if (file_str.endsWith('.fastq.gz') || file_str.endsWith('.fastq')) {
        // read2 exists for paired-end FASTQ?
        def read2File = replaceReadNumber(file_str)
        if (new File(read2File).exists()) {
            return 'fastq_paired_end'
        } else {
            return 'fastq_single_end'
        }
    } else {
        println "Unknown file type for $file_str. If fastq, check if _R1_ or .R1. patterns are used."
        return 'unknown'
    }
}


workflow create_placeholder { touchFile("${workflow.workDir}/placeholder.r2.fastq") }

workflow {
    create_placeholder()
    main:
        // placeholder for R2 file, can't be a random file as that would break nextflow's caching features
        // create the FILE here so it actually exists (touch)
        //touchFile("${workflow.workDir}/placeholder.r2.fastq")
        placeholder_r2 = file("${workflow.workDir}/placeholder.r2.fastq")
        
        // if reference is not indexed, index it.
        if (!file(params.path_to_genome_fasta).exists()) {
            println "Workflow failed: Genome file does not exist."
            System.exit(1)  // Exit with a custom status code
        }

        genome_index_ch = bwa_index()

        reads = Channel
          .fromPath(params.input_glob)
          .filter { !(it.name ==~ /^.*2\.fastq.*$/) }
          .map { input_file ->
            def fileType = detectFileType(input_file)
            def read1File = input_file
            def read2File = placeholder_r2.toString()
            if (fileType == 'fastq_paired_end') {
                read2File = replaceReadNumber(input_file.toString())
           }
            if (read1File.toString() == read2File) {
                log.error("Error: Detected paired-end file with read1: ${read1File} but no read2. What is different in the file name?")
                throw new IllegalStateException("Invalid paired-end file configuration")
            }
	        def library = read1File.baseName.replaceFirst(/.fastq|.fastq.gz|.bam/,"").replaceFirst(/_1|\.1|.R1/,"")
            return [params.email, library, read1File, read2File, fileType]
          }
          //.join(genome_index_ch)
        

        reads.view()


        println "Processing " + params.flowcell + "... => " + params.outputDir
        println "Cmd line: $workflow.commandLine"


        // files with few reads will be filtered out and used will get an email.
        checking_reads = enough_reads(reads)
        passed_reads = checking_reads
                       .filter { tuple -> tuple[5].text.contains('pass') }
                       .map { tuple -> tuple[0..4] }
        failed_reads = checking_reads.filter { tuple -> tuple[5].text.contains('fail') }.map { tuple -> tuple[5].text }
        send_email( failed_reads.collect() )
        

        // align and mark duplicates
        alignedReads = alignReads( passed_reads, genome_index_ch )
        markDup      = mergeAndMarkDuplicates( alignedReads.bam_files )
        extract      = methylDackel_extract( markDup.md_bams, genome_index_ch )
        mbias        = methylDackel_mbias( markDup.md_bams, genome_index_ch )

        // collect statistics
        gcbias       = gc_bias( markDup.md_bams, genome_index_ch )
        idxstats     = idx_stats( markDup.md_bams )
        flagstats    = flag_stats( markDup.md_bams )
        fastqc       = fastqc( markDup.md_bams )
        insertsize   = insert_size_metrics( markDup.md_bams ) 
        metrics      = picard_metrics( markDup.md_bams, genome_index_ch )
        mismatches   = tasmanian( markDup.md_bams, genome_index_ch )

        // Channels and processes that summarize all results

        // channel for internal summaries
        grouped_email_library = reads
	      .join( alignedReads.for_agg.groupTuple(by: [0, 1]), by: [0,1])
            .join( markDup.for_agg.groupTuple(by: [0,1]), by: [0,1] )
            .join( gcbias.for_agg.groupTuple(by: [0,1]), by: [0,1] )
            .join( idxstats.for_agg.groupTuple(by: [0,1]), by: [0,1] )
            .join( flagstats.for_agg.groupTuple(by: [0,1]), by: [0,1] )
            .join( fastqc.for_agg.groupTuple(by: [0,1]), by: [0,1] )
            .join( insertsize.for_agg.groupTuple(by: [0,1]), by: [0,1] )
            .join( mismatches.for_agg.groupTuple(by: [0,1]), by: [0,1] )
            .join( mbias.for_agg.groupTuple(by: [0,1]), by: [0,1] )
            .join( metrics.for_agg.groupTuple(by: [0,1]), by: [0,1] )

        if (params.enable_neb_agg.toString().toUpperCase() == "TRUE") {
            aggregate_emseq( grouped_email_library ) 
        }
       
        // channel for multiqc analysis
        all_results = grouped_email_library
         .join(insertsize.high_mapq_insert_size_metrics.groupTuple(by: [0, 1]), by: [0, 1])
         .map { items -> [items[0], items[7..-1]] }
         .groupTuple()
         .flatten()
         .toList()
         .map { items -> [items[0], items[7..-1]] }

        multiqc( all_results )
}
