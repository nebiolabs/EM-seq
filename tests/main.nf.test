nextflow_pipeline {

    name "Test EM-seq"
    script "main.nf"
    tag "pipeline"

    test("EM-seq workflow uBam") {
        when {
            params.ubam_dir = "$projectDir/tests/fixtures/ubam/"
            params.genome = 'test'
            params.email = 'foo@example.com'
            params.max_input_reads = 10000
            params.flowcell = "test_pipeline"
            params.enable_neb_agg = "false"

        }

        then {
            def bwameth = bam("${launchDir}/test_output/bwameth_align/emseq-test1.aln.bam").getStatistics()
            def bwameth2 = path("${launchDir}/test_output/bwameth_align/emseq-test2.aln.bam").exists()
            def fastqc = path("${launchDir}/test_output/stats/fastqc/emseq-test1.md_fastqc.html").exists()
            def fastqc2 = path("${launchDir}/test_output/stats/fastqc/emseq-test2.md_fastqc.html").exists()
            def fastp = path("${launchDir}/test_output/bwameth_align/emseq-test1.fastp.json").text.tokenize('\n')[1..33]
            def fastp2 = path("${launchDir}/test_output/bwameth_align/emseq-test2.fastp.json").exists()
            def markdup = bam("${launchDir}/test_output/markduped_bams/emseq-test1.md.bam").getStatistics()
            def markdup2 = path("${launchDir}/test_output/markduped_bams/emseq-test2.md.bam").exists()
            def insert_sizes = path("${launchDir}/test_output/stats/insert_size/emseq-test1.insertsize_metrics").text.tokenize('\n')[5..8]
            def insert_sizes2 = path("${launchDir}/test_output/stats/insert_size/emseq-test2.insertsize_metrics").exists()
            def idxstats = path("${launchDir}/test_output/stats/idxstats/emseq-test1.idxstat").text.tokenize('\n')
            def idxstats2 = path("${launchDir}/test_output/stats/idxstats/emseq-test2.idxstat").exists()
            def flagstats = path("${launchDir}/test_output/stats/flagstats/emseq-test1.flagstat").text.tokenize('\n')
            def flagstats2 = path("${launchDir}/test_output/stats/flagstats/emseq-test2.flagstat").exists()
            def gc_bias = path("${launchDir}/test_output/stats/gc_bias/emseq-test1.gc_metrics").text.tokenize('\n')[5..-1]
            def gc_bias2 = path("${launchDir}/test_output/stats/gc_bias/emseq-test2.gc_metrics").exists()
            def alignment_metrics = path("${launchDir}/test_output/stats/picard_alignment_metrics/emseq-test1.alignment_summary_metrics.txt").text.tokenize('\n')[5..8]
            def alignment_metrics2 = path("${launchDir}/test_output/stats/picard_alignment_metrics/emseq-test2.alignment_summary_metrics.txt").exists()
            def methyldackel_extract = path("${launchDir}/test_output/methylDackelExtracts/emseq-test1_CpG.methylKit.gz").md5
            def methyldackel_extract2 = path("${launchDir}/test_output/methylDackelExtracts/emseq-test2_CpG.methylKit.gz").exists()
            def mbias = path("${launchDir}/test_output/methylDackelExtracts/mbias/emseq-test1.combined_mbias.tsv").md5
            def mbias2 = path("${launchDir}/test_output/methylDackelExtracts/mbias/emseq-test2.combined_mbias.tsv").exists()
            def nonconverted = path("${launchDir}/test_output/bwameth_align/emseq-test1.nonconverted.tsv").text.tokenize('\n')
            def nonconverted2 = path("${launchDir}/test_output/bwameth_align/emseq-test2.nonconverted.tsv").exists()
            def intersection_summaries = path("${launchDir}/test_output/methylKit_intersections/all_summaries_combined.tsv").text.tokenize('\n')
            def methylkit_intersections = path("${launchDir}/test_output/methylKit_intersections/all_intersections_combined.tsv").text.tokenize('\n')
            def tasmanian = path("${launchDir}/test_output/stats/tasmanian/emseq-test1.csv").text.tokenize('\n')
            def tasmanian2 = path("${launchDir}/test_output/stats/tasmanian/emseq-test2.csv").exists()
            def multiqc = path("${launchDir}/test_output/EM-seq-Alignment-Summary-test_pipeline_multiqc_report.html").exists()

            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.trace,
                    ["bwameth", bwameth],
                    ["bwameth2", bwameth2],
                    ["fastqc", fastqc],
                    ["fastqc2", fastqc2],
                    ["fastp", fastp],
                    ["fastp2", fastp2],
                    ["markdup", markdup],
                    ["markdup2", markdup2],
                    ["insert_sizes", insert_sizes],
                    ["insert_sizes2", insert_sizes2],
                    ["idxstats", idxstats],
                    ["idxstats2", idxstats2],
                    ["flagstats", flagstats],
                    ["flagstats2", flagstats2],
                    ["gc_bias", gc_bias],
                    ["gc_bias2", gc_bias2],
                    ["alignment_metrics", alignment_metrics],
                    ["alignment_metrics2", alignment_metrics2],
                    ["methyldackel_extract", methyldackel_extract],
                    ["methyldackel_extract2", methyldackel_extract2],
                    ["mbias", mbias],
                    ["mbias2", mbias2],
                    ["nonconverted", nonconverted],
                    ["nonconverted2", nonconverted2],
                    ["intersection_summaries", intersection_summaries],
                    ["methylkit_intersections", methylkit_intersections],
                    ["tasmanian", tasmanian],
                    ["tasmanian2", tasmanian2],
                    ["multiqc", multiqc]
                ).match()
            }
        )
        }
    }
}
