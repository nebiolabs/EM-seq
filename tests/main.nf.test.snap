{
    "EM-seq workflow uBam": {
        "content": [
            {
                "tasksFailed": 0,
                "tasksCount": 33,
                "tasksSucceeded": 33
            },
            [
                "bwameth",
                {
                    "maxReadLength": 76,
                    "minReadLength": 31,
                    "meanReadLength": 74,
                    "maxQuality": 60,
                    "minQuality": 0,
                    "meanQuality": 18,
                    "readCount": 1801
                }
            ],
            [
                "fastqc",
                true
            ],
            [
                "fastp",
                [
                    "\t\"summary\": {",
                    "\t\t\"fastp_version\": \"0.23.4\",",
                    "\t\t\"sequencing\": \"paired end (76 cycles + 76 cycles)\",",
                    "\t\t\"before_filtering\": {",
                    "\t\t\t\"total_reads\":9534,",
                    "\t\t\t\"total_bases\":724584,",
                    "\t\t\t\"q20_bases\":681052,",
                    "\t\t\t\"q30_bases\":662196,",
                    "\t\t\t\"q20_rate\":0.939921,",
                    "\t\t\t\"q30_rate\":0.913898,",
                    "\t\t\t\"read1_mean_length\":76,",
                    "\t\t\t\"read2_mean_length\":76,",
                    "\t\t\t\"gc_content\":0.218668",
                    "\t\t},",
                    "\t\t\"after_filtering\": {",
                    "\t\t\t\"total_reads\":9526,",
                    "\t\t\t\"total_bases\":719174,",
                    "\t\t\t\"q20_bases\":676534,",
                    "\t\t\t\"q30_bases\":657910,",
                    "\t\t\t\"q20_rate\":0.94071,",
                    "\t\t\t\"q30_rate\":0.914813,",
                    "\t\t\t\"read1_mean_length\":75,",
                    "\t\t\t\"read2_mean_length\":75,",
                    "\t\t\t\"gc_content\":0.216799",
                    "\t\t}",
                    "\t},",
                    "\t\"filtering_result\": {",
                    "\t\t\"passed_filter_reads\": 9526,",
                    "\t\t\"low_quality_reads\": 0,",
                    "\t\t\"too_many_N_reads\": 0,",
                    "\t\t\"too_short_reads\": 8,",
                    "\t\t\"too_long_reads\": 0",
                    "\t},"
                ]
            ],
            [
                "markdup",
                {
                    "maxReadLength": 76,
                    "minReadLength": 31,
                    "meanReadLength": 74,
                    "maxQuality": 60,
                    "minQuality": 0,
                    "meanQuality": 18,
                    "readCount": 1801
                }
            ],
            [
                "markdup_log",
                true
            ],
            [
                "insert_sizes",
                true
            ],
            [
                "idxstats",
                [
                    "chr1\t284940\t1537\t264",
                    "*\t0\t0\t0"
                ]
            ],
            [
                "flagstats",
                true
            ],
            [
                "gc_bias",
                true
            ],
            [
                "alignment_metrics",
                true
            ],
            [
                "methyldackel_extract",
                true
            ],
            [
                "mbias",
                true
            ],
            [
                "nonconverted",
                [
                    "chr1\t0"
                ]
            ],
            [
                "intersection_summaries",
                true
            ],
            [
                "methylkit_intersections",
                true
            ],
            [
                "tasmanian",
                true
            ],
            [
                "multiqc",
                true
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.2"
        },
        "timestamp": "2025-08-25T12:47:45.62038133"
    }
}